include/heuristic/HeuristicMapper.hpp

```diff
@@ -275,28 +240,174 @@ class HeuristicMapper : public Mapper {
+   *
+   * @param node search node for which to recalculate the fixed cost
+   */
+  void recalculateFixedCostNonFidelity(Node& node);
 
   /**
-   * @brief calculates the heuristic cost for the following layers and saves it
-   * in the node as `lookaheadPenalty`
+   * @brief recalculates the gate-count-optimizing fixed cost of all reversals
+   * in the current mapping of a goal node or sets it to 0 otherwise
+   *
+   * @param layer index of current circuit layer
+   * @param node search node for which to recalculate the fixed cost
+   */
+  void recalculateFixedCostReversals(std::size_t layer, Node& node);
+
+  /**
+   * @brief calculates the heuristic cost of the current mapping in the node
+   * for some given layer and writes it to `Node::costHeur`, additionally
+   * `Node::validMapping` is set to true if all qubit pairs sharing a gate in
+   * the current layer are mapped next to each other
+   *
+   * @param layer index of current circuit layer
+   * @param node search node for which to calculate the heuristic cost
+   */
+  void updateHeuristicCost(std::size_t layer, Node& node);
+
+  /**
+   * @brief calculates the heuristic using `Heuristic::GateCountMaxDistance`
+   *
+   * @param layer index of current circuit layer
+   * @param node search node for which to calculate the heuristic cost
+   *
+   * @return heuristic cost
+   */
+  double heuristicGateCountMaxDistance(std::size_t layer, Node& node);
+
+  /**
+   * @brief calculates the heuristic using `Heuristic::GateCountSumDistance`
+   *
+   * @param layer index of current circuit layer
+   * @param node search node for which to calculate the heuristic cost
+   *
+   * @return heuristic cost
+   */
+  double heuristicGateCountSumDistance(std::size_t layer, Node& node);
+
+  /**
+   * @brief calculates the heuristic using
+   * `Heuristic::GateCountSumDistanceMinusSharedSwaps`
+   *
+   * @param layer index of current circuit layer
+   * @param node search node for which to calculate the heuristic cost
+   *
+   * @return heuristic cost
+   */
+  double heuristicGateCountSumDistanceMinusSharedSwaps(std::size_t layer,
+                                                       Node&       node);
+
+  /**
+   * @brief calculates the heuristic using
+   * `Heuristic::GateCountMaxDistanceOrSumDistanceMinusSharedSwaps`
+   *
+   * @param layer index of current circuit layer
+   * @param node search node for which to calculate the heuristic cost
+   *
+   * @return heuristic cost
+   */
+  double
+  heuristicGateCountMaxDistanceOrSumDistanceMinusSharedSwaps(std::size_t layer,
+                                                             Node&       node);
+
+  /**
+   * @brief calculates the heuristic using
+   * `Heuristic::FidelityBestLocation`
+   *
+   * @param layer index of current circuit layer
+   * @param node search node for which to calculate the heuristic cost
+   *
+   * @return heuristic cost
+   */
+  double heuristicFidelityBestLocation(std::size_t layer, Node& node);
+
```

Comment:
Another side comment/observation: I could be worth exploring whether we could turn the Heuristics from an enum to a proper class (hierarchy) so that methods like these could be bundled into a class and would not be so free-floating and loosely tied together. Same goes for the Lookahead strategies down below.
Definitely nothing for this PR though,

Reply:
Isn't an enum much more efficient in many regards (space-wise and runtime-wise in comparisons), since it's only an int in the backend?

Reply:
That kind of depends I guess. That Heuristics class would probably not have all that many member. So its memory footprint would probably be comparable. The comparison logic kind of depends on how it's implemented in the class, but could also be pretty comparable.
However, I am not really trying to push towards such a solution. I am not yet sure myself if it would amount to something useful. Let's keep it as it is for now, but keep the additional option in mind ☺️

Reply:
Alright, sure, let's keep it in mind☺️